src/rgw/rgw_zone_features.h

```diff
@@ -9,17 +9,25 @@
 
 #include <string>
 #include <boost/container/flat_set.hpp>
+#include "include/encoding.h"
 
 namespace rgw::zone_features {
 
+enum class feature_state {
+    disabled,
+    migrating,
+    enabled
```

Comment:
we start migration only fter the entire zonegroup is upgraded to the new code.
but while we do that we need to tesl lthe RGWs to not make any modifications to topic configuration.
this is a global zonegroup definition, because when we are at v1 they don't forward the requests to the master zone

Reply:
> this is a global zonegroup definition, because when we are at v1 they don't forward the requests to the master zone

we can fix that, right?

Reply:
instead of an explicit 'migrating' state in the zonegroup, consider this alternate definition:

the realm is "migrating" if
a) the notifications_v2 feature is enabled on all zonegroups, and
b) the v1 topics object still exists on the metadata master zone

on RadosStore startup, check whether we're in the "migrating" state on the metadata master zone. if so, try to acquire a cls_lock on the topics object. if successful, start a background thread or coroutine that runs the migration process under that lock, deleting the v1 topics object when it finishes

this way, the migration happens automatically on realm reload after the zonegroup feature is enabled. we don't need a separate state transition in the zonegroup after migration completes. we don't need to document or maintain extra radosgw-admin commands for migration

Reply:
this approach also means there's no gap after the v2 feature is enabled but before the migration starts where we might write v2 metadata before the same topic/notification gets migrated from v1 to v2. so the migration logic shouldn't need to deal with cases where v2 metadata already exists

Reply:
this sounds like a better aproach. couple of comments:
* "the notifications_v2 feature is enabled on all zonegroups" how can we verify that the feature is enabled for all zonegroups?
* @kchheda3 do you forward topic and notification configuration to the master zone for v1 as well?
* "start a background thread or coroutine" wouldn't it be better to actually block the rgw startup process during the migration? this will save us the need to explicitly reject topic ops that may arrive to other RGWs?

Reply:
> * @kchheda3 do you forward topic and notification configuration to the master zone for v1 as well?

@yuvalif, yes we do forward every request to `master zone` (we do not check v1 or v2 but forward all request to master)

Reply:
> @yuvalif, yes we do forward every request to `master zone` (we do not check v1 or v2 but forward all request to master)

`RGWPSCreateNotifOp::execute_v2()` does call `rgw_forward_request_to_master()`, but the v1 `execute()` does not. same with `RGWPSDeleteNotifOp`

Reply:
> > @yuvalif, yes we do forward every request to `master zone` (we do not check v1 or v2 but forward all request to master)
> 
> `RGWPSCreateNotifOp::execute_v2()` does call `rgw_forward_request_to_master()`, but the v1 `execute()` does not. same with `RGWPSDeleteNotifOp`

Oops, thanks @cbodley for pointing that
So the topic related operations are forwarded to primary, but notifications related are not. 
Shall I update the code to forward notification requests to master zone as well ?

Reply:
> > > @yuvalif, yes we do forward every request to `master zone` (we do not check v1 or v2 but forward all request to master)
> > 
> > 
> > `RGWPSCreateNotifOp::execute_v2()` does call `rgw_forward_request_to_master()`, but the v1 `execute()` does not. same with `RGWPSDeleteNotifOp`
> 
> Oops, thanks @cbodley for pointing that So the topic related operations are forwarded to primary, but notifications related are not. Shall I update the code to forward notification requests to master zone as well ?

yes, otherwise it would be difficult to reject the requests while we are in "migration" state - as only the master zone has knowledge about that

Reply:
> > > > @yuvalif, yes we do forward every request to `master zone` (we do not check v1 or v2 but forward all request to master)
> > > 
> > > 
> > > `RGWPSCreateNotifOp::execute_v2()` does call `rgw_forward_request_to_master()`, but the v1 `execute()` does not. same with `RGWPSDeleteNotifOp`
> > 
> > 
> > Oops, thanks @cbodley for pointing that So the topic related operations are forwarded to primary, but notifications related are not. Shall I update the code to forward notification requests to master zone as well ?
> 
> yes, otherwise it would be difficult to reject the requests while we are in "migration" state - as only the master zone has knowledge about that

updated the PR to forward to master zone.

Reply:
> for the case of migrating an existing cluster, when we enable v2, we can do the same for the topics on the master zone, but how would we migrate the topics that were created on other zones?

@yuvalif good question.. i'm not sure that we reasonably can do migration anywhere but the metadata master zone, since all metadata writes have to happen there ultimately

after upgrade/migration on one zone, the admin could potentially [promote a different zone](https://docs.ceph.com/en/latest/radosgw/multisite/#changing-the-metadata-master-zone) to metadata master. if that newly-promoted zone still has old v1 metadata sitting around, we'd trigger another migration. but how would we resolve migration conflicts? just ignore v1 topics that already have v2 metadata, and v1 notifications that have the v2 bucket xattr?

this failover process does need to work, but i'm skeptical that we want to require multiple failovers just to upgrade to squid. each failover risks data loss if the promoted zone wasn't caught up with metadata sync. maybe it's good enough to document that any topic/notification metadata on secondary zones should be manually copied to the master zone before enabling notifications-v2?

Reply:
> > for the case of migrating an existing cluster, when we enable v2, we can do the same for the topics on the master zone, but how would we migrate the topics that were created on other zones?
> 
> @yuvalif good question.. i'm not sure that we reasonably can do migration anywhere but the metadata master zone, since all metadata writes have to happen there ultimately
> 
> after upgrade/migration on one zone, the admin could potentially [promote a different zone](https://docs.ceph.com/en/latest/radosgw/multisite/#changing-the-metadata-master-zone) to metadata master. if that newly-promoted zone still has old v1 metadata sitting around, we'd trigger another migration. but how would we resolve migration conflicts? just ignore v1 topics that already have v2 metadata, and v1 notifications that have the v2 bucket xattr?
> 
> this failover process does need to work, but i'm skeptical that we want to require multiple failovers just to upgrade to squid. each failover risks data loss if the promoted zone wasn't caught up with metadata sync. maybe it's good enough to document that any topic/notification metadata on secondary zones should be manually copied to the master zone before enabling notifications-v2?

or we can stay with a radosgw-admin command solution?
the command will loop over all zones and migrate the topics for each one. we can figure out how to resolve conflicts in the commad, as long as what we do is documented, i don't see a big issue (the topics will probably be the same on all zones).

Reply:
how does the admin command help with other zones?

Reply:
can't we get the list of all zones with gateways?